text/chapter10.md

```diff
@@ -31,330 +36,372 @@ PureScript's FFI is designed to be very flexible. In practice, this means that d
-foreign import unsafeHead :: forall a. Array a -> a
-```
+We can instead return a `Maybe` value to handle this corner case.
 
-In our foreign JavaScript module, we can define `unsafeHead` as follows:
+It is tempting to write the following:
+```js
+exports.maybeHead = arr => {
+  if (arr.length) {
+    return Data_Maybe.Just.create(arr[0]);
+  } else {
+    return Data_Maybe.Nothing.value;
+  }
+}
+```
 
-```javascript
-exports.unsafeHead = function(arr) {
+```hs
+foreign import maybeHead :: forall a. Array a -> Maybe a
+```
+
+But calling these `Maybe` constructors directly in the FFI code isn't recommended as it makes the code brittle to changes in the code generator. Additionally, doing this can cause problems when using `purs bundle` for dead code elimination.
+
+The recommended approach is to add extra parameters to your FFI-defined function to accept the functions you need to call as arguments:
+
+```js
+exports.maybeHeadImpl = just => nothing => arr => {
   if (arr.length) {
-    return arr[0];
+    return just(arr[0]);
   } else {
-    throw new Error('unsafeHead: empty array');
+    return nothing;
   }
 };
 ```
 
-## Defining Foreign Types
+```hs
+foreign import maybeHeadImpl :: forall a. (forall x. x -> Maybe x) -> (forall x. Maybe x) -> Array a -> Maybe a
 
-Throwing an exception in the case of failure is less than ideal - idiomatic PureScript code uses the type system to represent side-effects such as missing values. One example of this approach is the `Maybe` type constructor. In this section, we will build another solution using the FFI.
+maybeHead :: forall a. Array a -> Maybe a
+maybeHead arr = maybeHeadImpl Just Nothing arr
+```
 
-Suppose we wanted to define a new type `Undefined a` whose representation at runtime was like that for the type `a`, but also allowing the `undefined` value.
+Note that we wrote:
+```hs
+forall a. (forall x. x -> Maybe x) -> (forall x. Maybe x) -> Array a -> Maybe a
+```
+and not:
+```hs
+forall a. ( a -> Maybe a) -> Maybe a -> Array a -> Maybe a
+```
+
+While both forms work, the latter is more vulnerable to unwanted inputs in place of `Just` and `Nothing`.
+For example, in the more vulnerable case we could call it as follows:
+```hs
+maybeHeadImpl (\_ -> Just 1000) (Just 1000) [1,2,3]
+```
+which returns `Just 1000` for any array input.
+This vulnerability is allowed because `(\_ -> Just 1000)` and `Just 1000` match the signatures of `(a -> Maybe a)` and `Maybe a` respectively when `a` is `Int` (based on input array).
+
+In the more secure type signature, even when `a` is determined to be `Int` based on the input array, we still need to provide valid functions matching the signatures involving `forall x`.
+The *only* option for `(forall x. Maybe x)` is `Nothing`, since a `Just` would assume a type for `x` and then no longer be valid for all `x`. The only options for `(forall x. x -> Maybe x)` are `Just` (our desired argument) and `(\_ -> Nothing)`, which is the only remaining vulnerability.
+
```

Comment:
Questioning whether the above distinction is necessary. The vulnerability is really just if there's a mistake in the adjacent wrapper, and the suggested approach doesn't eliminate all vulnerabilities. The suggested approach can't be applied to most other wrapping strategies, either.